### ISSUE TITLE
altroot for datasets

### ISSUE THREAD

**Issue by jplitza** [openzfs/zfs#10968]
### Describe the feature would like to see added to OpenZFS

A property like "altroot", but for datasets, that will be prepended to descendant datasets' mountpoints even when they have a non-inherited "mountpoint" property.

### How will this feature improve OpenZFS?

When sending backups of (multiple, and primarily root) ZFS pools onto other machines, the mountpoint property of the received datasets can screw up your mounts, because they will be mounted somewhere you did not anticipate.

Consider the dataset `rpool/var/lib/docker` on the host to be backed up, where obviously `rpool` has `mountpoint=/`. That dataset will be sent to `tank/backups/host1/rpool/var/lib/docker`. Even if `tank` has e.g. `altroot=/mnt` set, this dataset will be mounted at `/mnt/var/lib/docker`, colliding with any other such dataset on any other host that is backed up.

If the `tank` pool has no `altroot` set (e.g. because it is the root pool of the backup host), then the situation is event worse and the dataset will be mounted at `/var/lib/docker`, colliding even with the host filesystem hierarchy.

Of course, you can set `canmount=no`, but then you cannot mount the backups to look into them. Or you can unset the `mountpoint` of all received datasets, making restore a PITA.

### Additional context

I imagine this altroot to be non-escapable, in contrast to the `mountpoint` property, the same as the `altroot` of a pool. So if I have `tank/backups/host1/tank/backups/hostA` setting `altroot` as well, it would be appended to the altroot of `tank/backups/host1`.

**Comment by beren12** [comment]
After backing up a bsd box and getting bitten by this, it's a feature I would love to see. Set altroot on a dataset and nothing can be mounted outside of it, and it can preserve the backedup mountpoints.

**Comment by Shellcat-Zero** [comment]
I am very much interested and looking forward to this feature. It is very difficult, if not impossible, to install multiple bootable OSes in ZFS without having altroot for datasets.

**Comment by beren12** [comment]
If you are doing it yourself it’s not that bad, you need to set canmount=noauto and have your os basically in 1 dataset. Zfsbootmenu helps boot it as well.

**Comment by Shellcat-Zero** [comment]
> If you are doing it yourself it’s not that bad, you need to set canmount=noauto and have your os basically in 1 dataset. Zfsbootmenu helps boot it as well.

@beren12 My environments require grub, unfortunately, unless zfsbootmenu can be booted from grub.

**Comment by beren12** [comment]
Yes, it can. It's more of a zbm thing but I move my kernels out of /boot, install zbm, and run update-grub. I then mount /boot to /grub and put kernels back in /boot. This way you're real grub config will never get messed up, you an even keep /grub unmounted by default.
